pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/AbstractNode.java

```diff
@@ -431,22 +529,43 @@ public void setUserData(final Object userData) {
         this.userData = userData;
     }
 
+    /**
+     * @deprecated Not all nodes are based on tokens, and this is an implementation detail
+     */
+    @Deprecated
     public GenericToken jjtGetFirstToken() {
         return firstToken;
     }
 
+    /**
+     * @deprecated This is JJTree-specific and will be removed from this superclass.
```

Comment:
We can eg make it protected in AbstractJjtreeNode

Reply:
In fact, if you look at #2211, the method is only used by `JjtreeBuilder`, which is in the `ast.impl.javacc` package. So it could even be package-private in AbstractJjtreeNode.